packages/core/src/chopsticks-provider.ts

```diff
@@ -0,0 +1,287 @@
+}
+
+/**
+ * A provider for ApiPromise.
+ *
+ * Currectly only support browser environment.
+ */
+export class ChopsticksProvider implements ProviderInterface {
+  #isConnected = false
+  #eventemitter: EventEmitter
+  #isReadyPromise: Promise<void>
+  #endpoint: string
+  readonly stats?: ProviderStats
+  #subscriptions: Record<string, Subscription> = {}
+  #worker: Worker
+  #blockHash: string | undefined
+  #dbPath: string | undefined
+  #storageValues: StorageValues | undefined
+  #handlers: Record<string, Handler> = {}
+
+  constructor({ endpoint, blockHash, dbPath, storageValues }: ChopsticksProviderProps) {
+    if (!endpoint) {
+      throw new Error('ChopsticksProvider requires the upstream endpoint')
+    }
+    this.#endpoint = endpoint
+    this.#blockHash = blockHash
+    this.#dbPath = dbPath
+    this.#storageValues = storageValues
+
+    this.#eventemitter = new EventEmitter()
+
+    this.#isReadyPromise = new Promise((resolve, reject): void => {
+      this.#eventemitter.once('connected', (): void => {
+        logger.info('[Chopsticks provider] isReadyPromise: connected.')
+        resolve()
+      })
+      this.#eventemitter.once('error', reject)
+    })
+
+    const chopsticksWorker = new Worker(new URL('./chopsticks-worker', import.meta.url), { type: 'module' })
+    this.#worker = chopsticksWorker
+
+    this.connect()
+  }
+
+  get hasSubscriptions(): boolean {
+    return true
+  }
+
+  get isClonable(): boolean {
+    return true
+  }
+
+  get isConnected(): boolean {
+    return this.#isConnected
+  }
+
+  get isReady(): Promise<void> {
+    return this.#isReadyPromise
+  }
+
+  clone = (): ProviderInterface => {
+    return new ChopsticksProvider({ endpoint: this.#endpoint })
+  }
+
+  connect = async (): Promise<void> => {
+    if (this.#isConnected) {
+      return
+    }
+
+    this.#worker!.onmessage = this.#onWorkerMessage
+
+    this.#worker?.postMessage({
+      type: 'connect',
+      endpoint: this.#endpoint,
+      blockHash: this.#blockHash,
+      dbPath: this.#dbPath,
+      storageValues: this.#storageValues,
+    })
+  }
+
+  disconnect = async (): Promise<void> => {
+    this.#worker?.postMessage({ type: 'disconnect' })
+    this.#isConnected = false
+    this.#eventemitter.emit('disconnected')
+  }
+
+  on = (type: ProviderInterfaceEmitted, sub: ProviderInterfaceEmitCb): (() => void) => {
+    this.#eventemitter.on(type, sub)
+
+    return (): void => {
+      this.#eventemitter.removeListener(type, sub)
+    }
+  }
+
+  send = async <T>(
+    method: string,
+    params: unknown[],
+    _isCacheable?: boolean,
+    subscription?: SubscriptionHandler,
+  ): Promise<T> => {
+    return new Promise<T>((resolve, reject): void => {
+      try {
+        if (!this.isConnected || this.#worker === undefined) {
+          throw new Error('Api is not connected')
+        }
+
+        if (method !== 'system_health') {
+          logger.info('[Chopsticks provider] send', { method, params })
```

Comment:
why exclude `system_health`?
also could make this debug level

Reply:
got too many logs from system health in development, forgot to recover